argocd/structure_repository.go

```diff
@@ -64,13 +84,16 @@ func flattenRepository(repository *application.Repository, d *schema.ResourceDat
 		"name":                    repository.Name,
 		"project":                 repository.Project,
 		// TODO: in case of repositoryCredentials existence, will perma-diff
-		//"username":                repository.Username,
+		//"username":                		repository.Username,
 		// TODO: ArgoCD API does not return sensitive data!
-		//"password":                repository.Password,
-		//"ssh_private_key":         repository.SSHPrivateKey,
-		//"tls_client_cert_key":     repository.TLSClientCertKey,
-		"tls_client_cert_data": repository.TLSClientCertData,
-		"type":                 repository.Type,
+		//"password":                		repository.Password,
+		//"ssh_private_key":         		repository.SSHPrivateKey,
+		//"tls_client_cert_key":     		repository.TLSClientCertKey,
+		"tls_client_cert_data":          repository.TLSClientCertData,
+		"type":                          repository.Type,
+		"githubapp_id":                  convertInt64ToString(repository.GithubAppId),
+		"githubapp_installation_id":     convertInt64ToString(repository.GithubAppInstallationId),
```

Comment:
To answer your question in https://github.com/oboukili/terraform-provider-argocd/pull/235#issuecomment-1456441202... I would conditionally set these keys on the map based on the value of `repository.GithubAppId` and `repository.GithubAppInstallationId`. E.g. 

```go
if repository.GithubAppId > 0 {
	r["githubapp_id"] = convertInt64ToString(repository.GithubAppId)
}

if repository.GithubAppInstallationId != 0 {
	r["githubapp_installation_id"] = convertInt64ToString(repository.GithubAppInstallationId)
}
```

Reply:
Had a brief look but don't see the issue - it's weird :) Will have to do some playing around and will try to find the time to do so this weekend.

Reply:
Hi @philamente,

I managed to figure out what was happening here. This behaviour is a result of _shortcomings_ in the ArgoCD API. The `RepoCredsService` does not implement a `Get` method, so the provider [makes use](https://github.com/oboukili/terraform-provider-argocd/blob/124d87be91701f969512d2b855e868929c0245db/argocd/resource_argocd_repository_credentials.go#L79-L81) of the `ListRepositoryCredentials` endpoint to read data about existing repository credentials. This function only [returns](https://github.com/argoproj/argo-cd/blob/7be094f38d06859b594b98eb75c7c70d39b80b1e/server/repocreds/repocreds.go#L58-L61) the `url` and `username` and so the provider is unable to track drift between what is configured in Terraform versus the actual state of the underlying Kubernetes secret that ArgoCD stores the repo credentials in. As such, I have slimmed down the `flattenRepositoryCredentials` function to reflect this. I've also opened https://github.com/argoproj/argo-cd/issues/12823 to see if we can improve the situation in future.

Reply:
Also, as I went through the code, I addressed any other issues I noticed. I hope you don't mind me doing this; for your first contribution in Go, you did a great job, and I was hesitant about "denying" you further learning opportunities. However, given your last comment, I figured you wouldn't mind my expediting the process. I've kept the changes I made atomic, so feel free to take a look at the commits I added and let me know if anything is unclear or if you'd like me to explain any of them. Other than that, this is good to go now. So, once you let me know you're happy with the changes and, if there aren't any further changes from your side, I'll approve and merge.

Reply:
> Also, as I went through the code, I addressed any other issues I noticed. I hope you don't mind me doing this; for your first contribution in Go, you did a great job, and I was hesitant about "denying" you further learning opportunities. However, given your last comment, I figured you wouldn't mind my expediting the process. I've kept the changes I made atomic, so feel free to take a look at the commits I added and let me know if anything is unclear or if you'd like me to explain any of them. Other than that, this is good to go now. So, once you let me know you're happy with the changes and, if there aren't any further changes from your side, I'll approve and merge.

Hey, first of all thanks for digging into it and identifying the issue with the underlying ArgoCD API. Also I am really appreciating your input via single "topic" commits which indeed helped me to understand the missing/to be improved parts. I do not mind that at all!

--> Happy with it and we can go ahead with approval and merging

P.S.: I feel sorry for this akward "for" instead "if" copy pasta error in the flattening part of structure_repository. Shame on me I missed it there ;)

Reply:
> P.S.: I feel sorry for this akward "for" instead "if" copy pasta error in the flattening part of structure_repository. Shame on me I missed it there ;)

All good... We all make silly mistakes at times :smile: